### ISSUE TITLE
Hello, how should I configure offline static assets or offline external plugins when running via livereload?

### ISSUE THREAD

**Issue by jaywhj** [MartenBE/mkslides#27]
Hello, I found that when mkslides is running, not all static assets are copied to the built runtime environment, such as images or other files in the docs directory, and static assets configured in the `mkslides.yml` are not processed, so that there are files missing at runtime. For example:

```yaml
plugins:
  - extra_css:
      - assets/layout.css
  - extra_javascript:
      - ../plugin/xxx.js
```

I don't know if this is the original design. If don't copy all of them, you will need to handle the association judgment in many places separately, but there is a great risk of missing some files, such as the example above, and all tags with `src` attributes in HTML, which will be very troublesome to judge.

Here's how I'm handling it now, when running via static HTML slideshow files, I will manually configure any static asset files that aren't copied to the runtime environment to a relative directory outside of `docs`. But when running via livereload, how do I configure these offline static assets?

For the processing rules of static assets, I suggest that it is possible to design all files under `docs` (except `.md` files) to be copied to the built directory at one time? In this way, there is no need to judge one by one, whether running via static files or livereload, there will be no problem.

This is a minimal reproduction：

[test-mkslides.zip](https://github.com/user-attachments/files/18210771/test-mkslides.zip)

**Comment by jaywhj** [comment]
Yes, mkdocs is to copy all files under docs (except .md files) to the built directory. 

If there are files with the same name or conflicting files in the docs directory and the built directory, mkdocs' rule is that the conflicting files are not copied, the generated html files have a higher priority.

**Comment by markNZed** [comment]
This seems quite a limitation. For now could we just copy all the files at the start when using `mkslides serve` ? Or could we have an option to specify the dir that the server uses ? Thx

**Comment by MartenBE** [comment]
A rework for the folder logic is planned for the coming month, I'll probably add a folder/file mapping then to copy files over.

**Comment by noirbizarre** [comment]
Hi 👋🏼 

I just wanted to share that I have using for the past months a modified version of `mkslides`  to support this use case.
The principle is very basic, you just declare some static assets to be copied (src to dst mapping):
```yaml
static:
  themes: themes
  assets/favicon.png: favicon.png
```
They all will be copied to the destination path in the build root. Given they are sibling to the resulting slides, everything is accessible with relative URLs (aka. you can access ` themes/whatever/style.css` or `favicon.png` directly)

I just pushed it if you are interested: https://github.com/noirbizarre/mkslides/tree/feature/static
I did not submit a pull-request as it might not be complete and might lack some tests, but it has been perfectly working for me.
With this approach, automatic parsing and copy of themes and URLs become a bit redundant but it has benefits:
- you can publish things that are not embedded in slides
- you can have shared themes with live reload that embed assets
- it support any unparsed links (like data-background)
- it should work for reveal.js plugins (not tested)

Would you be interested in or is it too different from what you had in mind?
Meanwhile, fell free to pick whatever you want, those are very small changes.

**Comment by MartenBE** [comment]
@noirbizarre thanks for the interest!

Right now `mkslides` uses a top down approach: we parse the md for what it needs and then copy those files over. Unfortunately, this makes it difficult for people wanting to do custom things like including static folders or other content.
I am now drawing inspiration from `mkdocs` for a bottom up approach: copying everything in the folder, except md files which will be converted to html files. This should solve a lot of the problems mentioned here and allow for something similar to the `nav` option in `mkdocs` to describe a custom dir tree in the generated index.html.

It takes some effort, though. The whole logic needs a different approach and the logic isn't valid anymore for the generation of slides of a single markdown file, which I may deprecate in favour of the dir logic (which anybody will need as soon they want images or to include something else.

It should be coming following month :) In the mean time, you can always put in the PR of your changes. If it passes the tests, it could be merged.

**Comment by noirbizarre** [comment]
That's some very interesting reasoning you just shared! I totally get it, and that's clearly not a small task!
I am looking forward to seeing that refactoring and the next version 🤩 

Meanwhile, I submitted my changes as a PR: https://github.com/MartenBE/mkslides/pull/33

**Comment by MartenBE** [comment]
Progress is tracked in https://github.com/MartenBE/mkslides/pull/34

**Comment by MartenBE** [comment]
I am currently through the hardest part of the refactor for `1.1.0`. If you would like to see if it fits your use case and solves the problems described here, please check out https://github.com/MartenBE/mkslides/pull/34/commits/4e36894734d3ec12ddcd4de1f068226f6416557f.

As you can see from the stats of https://github.com/MartenBE/mkslides/pull/34, it has become a large refactor which is why it is taking so much time.

**Comment by MartenBE** [comment]
It took a long time as it was way more complex than I anticipated, but the rewrite of the folder logic is finally done! As this was a mayor change and took a lot of refactoring, the version tag is bumped to version 2.0.0. Try it out!
https://github.com/MartenBE/mkslides/releases/tag/2.0.0
